// kysely/dist/esm/operation-node/delete-query-node.js
/// <reference types="./delete-query-node.d.ts" />
import { freeze } from '../util/object-utils.js';
import { FromNode } from './from-node.js';
import { OrderByNode } from './order-by-node.js';
/**
 * @internal
 */
export const DeleteQueryNode = freeze({
    is(node) {
        return node.kind === 'DeleteQueryNode';
    },
    create(fromItem, withNode) {
        return freeze({
            kind: 'DeleteQueryNode',
            from: FromNode.create([fromItem]),
            ...(withNode && { with: withNode }),
        });
    },
    cloneWithOrderByItem(deleteNode, item) {
        return freeze({
            ...deleteNode,
            orderBy: deleteNode.orderBy
                ? OrderByNode.cloneWithItem(deleteNode.orderBy, item)
                : OrderByNode.create(item),
        });
    },
    cloneWithLimit(deleteNode, limit) {
        return freeze({
            ...deleteNode,
            limit,
        });
    },
    cloneWithExplain(deleteNode, explain) {
        return freeze({
            ...deleteNode,
            explain,
        });
    },
});
