// kysely/dist/esm/operation-node/on-conflict-node.js
/// <reference types="./on-conflict-node.d.ts" />
import { freeze } from '../util/object-utils.js';
import { WhereNode } from './where-node.js';
/**
 * @internal
 */
export const OnConflictNode = freeze({
    is(node) {
        return node.kind === 'OnConflictNode';
    },
    create() {
        return freeze({
            kind: 'OnConflictNode',
        });
    },
    cloneWith(node, props) {
        return freeze({
            ...node,
            ...props,
        });
    },
    cloneWithIndexWhere(node, filter) {
        return freeze({
            ...node,
            indexWhere: node.indexWhere
                ? WhereNode.cloneWithFilter(node.indexWhere, 'And', filter)
                : WhereNode.create(filter),
        });
    },
    cloneWithIndexOrWhere(node, filter) {
        return freeze({
            ...node,
            indexWhere: node.indexWhere
                ? WhereNode.cloneWithFilter(node.indexWhere, 'Or', filter)
                : WhereNode.create(filter),
        });
    },
    cloneWithUpdateWhere(node, filter) {
        return freeze({
            ...node,
            updateWhere: node.updateWhere
                ? WhereNode.cloneWithFilter(node.updateWhere, 'And', filter)
                : WhereNode.create(filter),
        });
    },
    cloneWithUpdateOrWhere(node, filter) {
        return freeze({
            ...node,
            updateWhere: node.updateWhere
                ? WhereNode.cloneWithFilter(node.updateWhere, 'Or', filter)
                : WhereNode.create(filter),
        });
    },
});
