// kysely/dist/esm/operation-node/query-node.js
/// <reference types="./query-node.d.ts" />
import { InsertQueryNode } from './insert-query-node.js';
import { SelectQueryNode } from './select-query-node.js';
import { UpdateQueryNode } from './update-query-node.js';
import { DeleteQueryNode } from './delete-query-node.js';
import { WhereNode } from './where-node.js';
import { freeze } from '../util/object-utils.js';
import { ReturningNode } from './returning-node.js';
/**
 * @internal
 */
export const QueryNode = freeze({
    is(node) {
        return (DeleteQueryNode.is(node) ||
            InsertQueryNode.is(node) ||
            UpdateQueryNode.is(node) ||
            SelectQueryNode.is(node));
    },
    cloneWithWhere(node, filter) {
        return freeze({
            ...node,
            where: node.where
                ? WhereNode.cloneWithFilter(node.where, 'And', filter)
                : WhereNode.create(filter),
        });
    },
    cloneWithOrWhere(node, filter) {
        return freeze({
            ...node,
            where: node.where
                ? WhereNode.cloneWithFilter(node.where, 'Or', filter)
                : WhereNode.create(filter),
        });
    },
    cloneWithJoin(node, join) {
        return freeze({
            ...node,
            joins: node.joins ? freeze([...node.joins, join]) : freeze([join]),
        });
    },
    cloneWithReturning(node, selections) {
        return freeze({
            ...node,
            returning: node.returning
                ? ReturningNode.cloneWithSelections(node.returning, selections)
                : ReturningNode.create(selections),
        });
    },
});
