// kysely/dist/esm/operation-node/select-query-node.js
/// <reference types="./select-query-node.d.ts" />
import { freeze } from '../util/object-utils.js';
import { FromNode } from './from-node.js';
import { GroupByNode } from './group-by-node.js';
import { HavingNode } from './having-node.js';
import { OrderByNode } from './order-by-node.js';
/**
 * @internal
 */
export const SelectQueryNode = freeze({
    is(node) {
        return node.kind === 'SelectQueryNode';
    },
    create(fromItems, withNode) {
        return freeze({
            kind: 'SelectQueryNode',
            from: FromNode.create(fromItems),
            ...(withNode && { with: withNode }),
        });
    },
    cloneWithSelections(select, selections) {
        return freeze({
            ...select,
            selections: select.selections
                ? freeze([...select.selections, ...selections])
                : freeze(selections),
        });
    },
    cloneWithDistinctOnSelections(select, selections) {
        return freeze({
            ...select,
            distinctOnSelections: select.distinctOnSelections
                ? freeze([...select.distinctOnSelections, ...selections])
                : freeze(selections),
        });
    },
    cloneWithFrontModifier(select, modifier) {
        return freeze({
            ...select,
            frontModifiers: select.frontModifiers
                ? freeze([...select.frontModifiers, modifier])
                : freeze([modifier]),
        });
    },
    cloneWithEndModifier(select, modifier) {
        return freeze({
            ...select,
            endModifiers: select.endModifiers
                ? freeze([...select.endModifiers, modifier])
                : freeze([modifier]),
        });
    },
    cloneWithOrderByItem(selectNode, item) {
        return freeze({
            ...selectNode,
            orderBy: selectNode.orderBy
                ? OrderByNode.cloneWithItem(selectNode.orderBy, item)
                : OrderByNode.create(item),
        });
    },
    cloneWithGroupByItems(selectNode, items) {
        return freeze({
            ...selectNode,
            groupBy: selectNode.groupBy
                ? GroupByNode.cloneWithItems(selectNode.groupBy, items)
                : GroupByNode.create(items),
        });
    },
    cloneWithLimit(selectNode, limit) {
        return freeze({
            ...selectNode,
            limit,
        });
    },
    cloneWithOffset(selectNode, offset) {
        return freeze({
            ...selectNode,
            offset,
        });
    },
    cloneWithHaving(selectNode, filter) {
        return freeze({
            ...selectNode,
            having: selectNode.having
                ? HavingNode.cloneWithFilter(selectNode.having, 'And', filter)
                : HavingNode.create(filter),
        });
    },
    cloneWithOrHaving(selectNode, filter) {
        return freeze({
            ...selectNode,
            having: selectNode.having
                ? HavingNode.cloneWithFilter(selectNode.having, 'Or', filter)
                : HavingNode.create(filter),
        });
    },
    cloneWithSetOperation(selectNode, setOperation) {
        return freeze({
            ...selectNode,
            setOperations: selectNode.setOperations
                ? freeze([...selectNode.setOperations, setOperation])
                : freeze([setOperation]),
        });
    },
    cloneWithExplain(selectNode, explain) {
        return freeze({
            ...selectNode,
            explain,
        });
    },
});
