// kysely/dist/cjs/schema/alter-table-builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AlterTableDropConstraintBuilder = exports.AlterTableAddForeignKeyConstraintBuilder = exports.AlterTableModifyColumnBuilder = exports.AlterTableAddColumnBuilder = exports.AlterTableExecutor = exports.AlterColumnBuilder = exports.AlterTableBuilder = void 0;
const add_column_node_js_1 = require("../operation-node/add-column-node.js");
const alter_column_node_js_1 = require("../operation-node/alter-column-node.js");
const alter_table_node_js_1 = require("../operation-node/alter-table-node.js");
const column_definition_node_js_1 = require("../operation-node/column-definition-node.js");
const data_type_node_js_1 = require("../operation-node/data-type-node.js");
const drop_column_node_js_1 = require("../operation-node/drop-column-node.js");
const identifier_node_js_1 = require("../operation-node/identifier-node.js");
const rename_column_node_js_1 = require("../operation-node/rename-column-node.js");
const object_utils_js_1 = require("../util/object-utils.js");
const prevent_await_js_1 = require("../util/prevent-await.js");
const column_definition_builder_js_1 = require("./column-definition-builder.js");
const modify_column_node_js_1 = require("../operation-node/modify-column-node.js");
const data_type_parser_js_1 = require("../parser/data-type-parser.js");
const foreign_key_constraint_builder_js_1 = require("./foreign-key-constraint-builder.js");
const add_constraint_node_js_1 = require("../operation-node/add-constraint-node.js");
const unique_constraint_node_js_1 = require("../operation-node/unique-constraint-node.js");
const check_constraint_node_js_1 = require("../operation-node/check-constraint-node.js");
const foreign_key_constraint_node_js_1 = require("../operation-node/foreign-key-constraint-node.js");
const column_node_js_1 = require("../operation-node/column-node.js");
const default_value_parser_js_1 = require("../parser/default-value-parser.js");
const table_parser_js_1 = require("../parser/table-parser.js");
const drop_constraint_node_js_1 = require("../operation-node/drop-constraint-node.js");
/**
 * This builder can be used to create a `alter table` query.
 */
class AlterTableBuilder {
    #props;
    constructor(props) {
        this.#props = (0, object_utils_js_1.freeze)(props);
    }
    renameTo(newTableName) {
        return new AlterTableExecutor({
            ...this.#props,
            alterTableNode: alter_table_node_js_1.AlterTableNode.cloneWith(this.#props.alterTableNode, {
                renameTo: (0, table_parser_js_1.parseTable)(newTableName),
            }),
        });
    }
    setSchema(newSchema) {
        return new AlterTableExecutor({
            ...this.#props,
            alterTableNode: alter_table_node_js_1.AlterTableNode.cloneWith(this.#props.alterTableNode, {
                setSchema: identifier_node_js_1.IdentifierNode.create(newSchema),
            }),
        });
    }
    alterColumn(column) {
        return new AlterColumnBuilder({
            ...this.#props,
            alterColumnNode: alter_column_node_js_1.AlterColumnNode.create(column),
        });
    }
    dropColumn(column) {
        return new AlterTableExecutor({
            ...this.#props,
            alterTableNode: alter_table_node_js_1.AlterTableNode.cloneWith(this.#props.alterTableNode, {
                dropColumn: drop_column_node_js_1.DropColumnNode.create(column),
            }),
        });
    }
    renameColumn(column, newColumn) {
        return new AlterTableExecutor({
            ...this.#props,
            alterTableNode: alter_table_node_js_1.AlterTableNode.cloneWith(this.#props.alterTableNode, {
                renameColumn: rename_column_node_js_1.RenameColumnNode.create(column, newColumn),
            }),
        });
    }
    /**
     * See {@link CreateTableBuilder.addColumn}
     */
    addColumn(columnName, dataType, build = object_utils_js_1.noop) {
        return build(new AlterTableAddColumnBuilder({
            ...this.#props,
            columnBuilder: new column_definition_builder_js_1.ColumnDefinitionBuilder(column_definition_node_js_1.ColumnDefinitionNode.create(columnName, (0, data_type_parser_js_1.parseDataTypeExpression)(dataType))),
        }));
    }
    /**
     * Creates an `alter table modify column` query. The `modify column` statement
     * is only implemeted by MySQL and oracle AFAIK. On other databases you
     * should use the `alterColumn` method.
     */
    modifyColumn(columnName, dataType) {
        return new AlterTableModifyColumnBuilder({
            ...this.#props,
            columnBuilder: new column_definition_builder_js_1.ColumnDefinitionBuilder(column_definition_node_js_1.ColumnDefinitionNode.create(columnName, (0, data_type_parser_js_1.parseDataTypeExpression)(dataType))),
        });
    }
    /**
     * See {@link CreateTableBuilder.addUniqueConstraint}
     */
    addUniqueConstraint(constraintName, columns) {
        return new AlterTableExecutor({
            ...this.#props,
            alterTableNode: alter_table_node_js_1.AlterTableNode.cloneWith(this.#props.alterTableNode, {
                addConstraint: add_constraint_node_js_1.AddConstraintNode.create(unique_constraint_node_js_1.UniqueConstraintNode.create(columns, constraintName)),
            }),
        });
    }
    /**
     * See {@link CreateTableBuilder.addCheckConstraint}
     */
    addCheckConstraint(constraintName, checkExpression) {
        return new AlterTableExecutor({
            ...this.#props,
            alterTableNode: alter_table_node_js_1.AlterTableNode.cloneWith(this.#props.alterTableNode, {
                addConstraint: add_constraint_node_js_1.AddConstraintNode.create(check_constraint_node_js_1.CheckConstraintNode.create(checkExpression.toOperationNode(), constraintName)),
            }),
        });
    }
    /**
     * See {@link CreateTableBuilder.addForeignKeyConstraint}
     *
     * Unlike {@link CreateTableBuilder.addForeignKeyConstraint} this method returns
     * the constraint builder and doesn't take a callback as the last argument. This
     * is because you can only add one column per `ALTER TABLE` query.
     */
    addForeignKeyConstraint(constraintName, columns, targetTable, targetColumns) {
        return new AlterTableAddForeignKeyConstraintBuilder({
            ...this.#props,
            constraintBuilder: new foreign_key_constraint_builder_js_1.ForeignKeyConstraintBuilder(foreign_key_constraint_node_js_1.ForeignKeyConstraintNode.create(columns.map(column_node_js_1.ColumnNode.create), (0, table_parser_js_1.parseTable)(targetTable), targetColumns.map(column_node_js_1.ColumnNode.create), constraintName)),
        });
    }
    dropConstraint(constraintName) {
        return new AlterTableDropConstraintBuilder({
            ...this.#props,
            alterTableNode: alter_table_node_js_1.AlterTableNode.cloneWith(this.#props.alterTableNode, {
                dropConstraint: drop_constraint_node_js_1.DropConstraintNode.create(constraintName),
            }),
        });
    }
}
exports.AlterTableBuilder = AlterTableBuilder;
class AlterColumnBuilder {
    #props;
    constructor(props) {
        this.#props = (0, object_utils_js_1.freeze)(props);
    }
    setDataType(dataType) {
        return new AlterTableExecutor({
            ...this.#props,
            alterTableNode: alter_table_node_js_1.AlterTableNode.cloneWith(this.#props.alterTableNode, {
                alterColumn: alter_column_node_js_1.AlterColumnNode.cloneWith(this.#props.alterColumnNode, {
                    dataType: data_type_node_js_1.DataTypeNode.create(dataType),
                }),
            }),
        });
    }
    setDefault(value) {
        return new AlterTableExecutor({
            ...this.#props,
            alterTableNode: alter_table_node_js_1.AlterTableNode.cloneWith(this.#props.alterTableNode, {
                alterColumn: alter_column_node_js_1.AlterColumnNode.cloneWith(this.#props.alterColumnNode, {
                    setDefault: (0, default_value_parser_js_1.parseDefaultValueExpression)(value),
                }),
            }),
        });
    }
    dropDefault() {
        return new AlterTableExecutor({
            ...this.#props,
            alterTableNode: alter_table_node_js_1.AlterTableNode.cloneWith(this.#props.alterTableNode, {
                alterColumn: alter_column_node_js_1.AlterColumnNode.cloneWith(this.#props.alterColumnNode, {
                    dropDefault: true,
                }),
            }),
        });
    }
    setNotNull() {
        return new AlterTableExecutor({
            ...this.#props,
            alterTableNode: alter_table_node_js_1.AlterTableNode.cloneWith(this.#props.alterTableNode, {
                alterColumn: alter_column_node_js_1.AlterColumnNode.cloneWith(this.#props.alterColumnNode, {
                    setNotNull: true,
                }),
            }),
        });
    }
    dropNotNull() {
        return new AlterTableExecutor({
            ...this.#props,
            alterTableNode: alter_table_node_js_1.AlterTableNode.cloneWith(this.#props.alterTableNode, {
                alterColumn: alter_column_node_js_1.AlterColumnNode.cloneWith(this.#props.alterColumnNode, {
                    dropNotNull: true,
                }),
            }),
        });
    }
}
exports.AlterColumnBuilder = AlterColumnBuilder;
class AlterTableExecutor {
    #props;
    constructor(props) {
        this.#props = (0, object_utils_js_1.freeze)(props);
    }
    toOperationNode() {
        return this.#props.executor.transformQuery(this.#props.alterTableNode, this.#props.queryId);
    }
    compile() {
        return this.#props.executor.compileQuery(this.toOperationNode(), this.#props.queryId);
    }
    async execute() {
        await this.#props.executor.executeQuery(this.compile(), this.#props.queryId);
    }
}
exports.AlterTableExecutor = AlterTableExecutor;
class AlterTableAddColumnBuilder {
    #props;
    constructor(props) {
        this.#props = (0, object_utils_js_1.freeze)(props);
    }
    autoIncrement() {
        return new AlterTableAddColumnBuilder({
            ...this.#props,
            columnBuilder: this.#props.columnBuilder.autoIncrement(),
        });
    }
    primaryKey() {
        return new AlterTableAddColumnBuilder({
            ...this.#props,
            columnBuilder: this.#props.columnBuilder.primaryKey(),
        });
    }
    references(ref) {
        return new AlterTableAddColumnBuilder({
            ...this.#props,
            columnBuilder: this.#props.columnBuilder.references(ref),
        });
    }
    onDelete(onDelete) {
        return new AlterTableAddColumnBuilder({
            ...this.#props,
            columnBuilder: this.#props.columnBuilder.onDelete(onDelete),
        });
    }
    onUpdate(onDelete) {
        return new AlterTableAddColumnBuilder({
            ...this.#props,
            columnBuilder: this.#props.columnBuilder.onUpdate(onDelete),
        });
    }
    unique() {
        return new AlterTableAddColumnBuilder({
            ...this.#props,
            columnBuilder: this.#props.columnBuilder.unique(),
        });
    }
    notNull() {
        return new AlterTableAddColumnBuilder({
            ...this.#props,
            columnBuilder: this.#props.columnBuilder.notNull(),
        });
    }
    unsigned() {
        return new AlterTableAddColumnBuilder({
            ...this.#props,
            columnBuilder: this.#props.columnBuilder.unsigned(),
        });
    }
    defaultTo(value) {
        return new AlterTableAddColumnBuilder({
            ...this.#props,
            columnBuilder: this.#props.columnBuilder.defaultTo(value),
        });
    }
    check(expression) {
        return new AlterTableAddColumnBuilder({
            ...this.#props,
            columnBuilder: this.#props.columnBuilder.check(expression),
        });
    }
    generatedAlwaysAs(expression) {
        return new AlterTableAddColumnBuilder({
            ...this.#props,
            columnBuilder: this.#props.columnBuilder.generatedAlwaysAs(expression),
        });
    }
    generatedAlwaysAsIdentity() {
        return new AlterTableAddColumnBuilder({
            ...this.#props,
            columnBuilder: this.#props.columnBuilder.generatedAlwaysAsIdentity(),
        });
    }
    generatedByDefaultAsIdentity() {
        return new AlterTableAddColumnBuilder({
            ...this.#props,
            columnBuilder: this.#props.columnBuilder.generatedByDefaultAsIdentity(),
        });
    }
    stored() {
        return new AlterTableAddColumnBuilder({
            ...this.#props,
            columnBuilder: this.#props.columnBuilder.stored(),
        });
    }
    toOperationNode() {
        return this.#props.executor.transformQuery(alter_table_node_js_1.AlterTableNode.cloneWith(this.#props.alterTableNode, {
            addColumn: add_column_node_js_1.AddColumnNode.create(this.#props.columnBuilder.toOperationNode()),
        }), this.#props.queryId);
    }
    compile() {
        return this.#props.executor.compileQuery(this.toOperationNode(), this.#props.queryId);
    }
    async execute() {
        await this.#props.executor.executeQuery(this.compile(), this.#props.queryId);
    }
}
exports.AlterTableAddColumnBuilder = AlterTableAddColumnBuilder;
class AlterTableModifyColumnBuilder {
    #props;
    constructor(props) {
        this.#props = (0, object_utils_js_1.freeze)(props);
    }
    autoIncrement() {
        return new AlterTableModifyColumnBuilder({
            ...this.#props,
            columnBuilder: this.#props.columnBuilder.autoIncrement(),
        });
    }
    primaryKey() {
        return new AlterTableModifyColumnBuilder({
            ...this.#props,
            columnBuilder: this.#props.columnBuilder.primaryKey(),
        });
    }
    references(ref) {
        return new AlterTableModifyColumnBuilder({
            ...this.#props,
            columnBuilder: this.#props.columnBuilder.references(ref),
        });
    }
    onDelete(onDelete) {
        return new AlterTableModifyColumnBuilder({
            ...this.#props,
            columnBuilder: this.#props.columnBuilder.onDelete(onDelete),
        });
    }
    onUpdate(onUpdate) {
        return new AlterTableModifyColumnBuilder({
            ...this.#props,
            columnBuilder: this.#props.columnBuilder.onUpdate(onUpdate),
        });
    }
    unique() {
        return new AlterTableModifyColumnBuilder({
            ...this.#props,
            columnBuilder: this.#props.columnBuilder.unique(),
        });
    }
    notNull() {
        return new AlterTableModifyColumnBuilder({
            ...this.#props,
            columnBuilder: this.#props.columnBuilder.notNull(),
        });
    }
    unsigned() {
        return new AlterTableModifyColumnBuilder({
            ...this.#props,
            columnBuilder: this.#props.columnBuilder.unsigned(),
        });
    }
    defaultTo(value) {
        return new AlterTableModifyColumnBuilder({
            ...this.#props,
            columnBuilder: this.#props.columnBuilder.defaultTo(value),
        });
    }
    check(expression) {
        return new AlterTableModifyColumnBuilder({
            ...this.#props,
            columnBuilder: this.#props.columnBuilder.check(expression),
        });
    }
    generatedAlwaysAs(expression) {
        return new AlterTableModifyColumnBuilder({
            ...this.#props,
            columnBuilder: this.#props.columnBuilder.generatedAlwaysAs(expression),
        });
    }
    generatedAlwaysAsIdentity() {
        return new AlterTableModifyColumnBuilder({
            ...this.#props,
            columnBuilder: this.#props.columnBuilder.generatedAlwaysAsIdentity(),
        });
    }
    generatedByDefaultAsIdentity() {
        return new AlterTableModifyColumnBuilder({
            ...this.#props,
            columnBuilder: this.#props.columnBuilder.generatedByDefaultAsIdentity(),
        });
    }
    stored() {
        return new AlterTableModifyColumnBuilder({
            ...this.#props,
            columnBuilder: this.#props.columnBuilder.stored(),
        });
    }
    toOperationNode() {
        return this.#props.executor.transformQuery(alter_table_node_js_1.AlterTableNode.cloneWith(this.#props.alterTableNode, {
            modifyColumn: modify_column_node_js_1.ModifyColumnNode.create(this.#props.columnBuilder.toOperationNode()),
        }), this.#props.queryId);
    }
    compile() {
        return this.#props.executor.compileQuery(this.toOperationNode(), this.#props.queryId);
    }
    async execute() {
        await this.#props.executor.executeQuery(this.compile(), this.#props.queryId);
    }
}
exports.AlterTableModifyColumnBuilder = AlterTableModifyColumnBuilder;
class AlterTableAddForeignKeyConstraintBuilder {
    #props;
    constructor(props) {
        this.#props = (0, object_utils_js_1.freeze)(props);
    }
    onDelete(onDelete) {
        return new AlterTableAddForeignKeyConstraintBuilder({
            ...this.#props,
            constraintBuilder: this.#props.constraintBuilder.onDelete(onDelete),
        });
    }
    onUpdate(onUpdate) {
        return new AlterTableAddForeignKeyConstraintBuilder({
            ...this.#props,
            constraintBuilder: this.#props.constraintBuilder.onUpdate(onUpdate),
        });
    }
    toOperationNode() {
        return this.#props.executor.transformQuery(alter_table_node_js_1.AlterTableNode.cloneWith(this.#props.alterTableNode, {
            addConstraint: add_constraint_node_js_1.AddConstraintNode.create(this.#props.constraintBuilder.toOperationNode()),
        }), this.#props.queryId);
    }
    compile() {
        return this.#props.executor.compileQuery(this.toOperationNode(), this.#props.queryId);
    }
    async execute() {
        await this.#props.executor.executeQuery(this.compile(), this.#props.queryId);
    }
}
exports.AlterTableAddForeignKeyConstraintBuilder = AlterTableAddForeignKeyConstraintBuilder;
class AlterTableDropConstraintBuilder {
    #props;
    constructor(props) {
        this.#props = (0, object_utils_js_1.freeze)(props);
    }
    ifExists() {
        return new AlterTableDropConstraintBuilder({
            ...this.#props,
            alterTableNode: alter_table_node_js_1.AlterTableNode.cloneWith(this.#props.alterTableNode, {
                dropConstraint: drop_constraint_node_js_1.DropConstraintNode.cloneWith(this.#props.alterTableNode.dropConstraint, {
                    ifExists: true,
                }),
            }),
        });
    }
    cascade() {
        return new AlterTableDropConstraintBuilder({
            ...this.#props,
            alterTableNode: alter_table_node_js_1.AlterTableNode.cloneWith(this.#props.alterTableNode, {
                dropConstraint: drop_constraint_node_js_1.DropConstraintNode.cloneWith(this.#props.alterTableNode.dropConstraint, {
                    modifier: 'cascade',
                }),
            }),
        });
    }
    restrict() {
        return new AlterTableDropConstraintBuilder({
            ...this.#props,
            alterTableNode: alter_table_node_js_1.AlterTableNode.cloneWith(this.#props.alterTableNode, {
                dropConstraint: drop_constraint_node_js_1.DropConstraintNode.cloneWith(this.#props.alterTableNode.dropConstraint, {
                    modifier: 'restrict',
                }),
            }),
        });
    }
    toOperationNode() {
        return this.#props.executor.transformQuery(this.#props.alterTableNode, this.#props.queryId);
    }
    compile() {
        return this.#props.executor.compileQuery(this.toOperationNode(), this.#props.queryId);
    }
    async execute() {
        await this.#props.executor.executeQuery(this.compile(), this.#props.queryId);
    }
}
exports.AlterTableDropConstraintBuilder = AlterTableDropConstraintBuilder;
(0, prevent_await_js_1.preventAwait)(AlterTableBuilder, "don't await AlterTableBuilder instances");
(0, prevent_await_js_1.preventAwait)(AlterColumnBuilder, "don't await AlterColumnBuilder instances");
(0, prevent_await_js_1.preventAwait)(AlterTableExecutor, "don't await AlterTableExecutor instances directly. To execute the query you need to call `execute`");
(0, prevent_await_js_1.preventAwait)(AlterTableAddColumnBuilder, "don't await AlterTableAddColumnBuilder instances directly. To execute the query you need to call `execute`");
(0, prevent_await_js_1.preventAwait)(AlterTableModifyColumnBuilder, "don't await AlterTableModifyColumnBuilder instances directly. To execute the query you need to call `execute`");
(0, prevent_await_js_1.preventAwait)(AlterTableAddForeignKeyConstraintBuilder, "don't await AlterTableAddForeignKeyConstraintBuilder instances directly. To execute the query you need to call `execute`");
(0, prevent_await_js_1.preventAwait)(AlterTableDropConstraintBuilder, "don't await AlterTableDropConstraintBuilder instances directly. To execute the query you need to call `execute`");
