// kysely/dist/cjs/schema/drop-table-builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DropTableBuilder = void 0;
const drop_table_node_js_1 = require("../operation-node/drop-table-node.js");
const prevent_await_js_1 = require("../util/prevent-await.js");
const object_utils_js_1 = require("../util/object-utils.js");
class DropTableBuilder {
    #props;
    constructor(props) {
        this.#props = (0, object_utils_js_1.freeze)(props);
    }
    ifExists() {
        return new DropTableBuilder({
            ...this.#props,
            dropTableNode: drop_table_node_js_1.DropTableNode.cloneWith(this.#props.dropTableNode, {
                ifExists: true,
            }),
        });
    }
    cascade() {
        return new DropTableBuilder({
            ...this.#props,
            dropTableNode: drop_table_node_js_1.DropTableNode.cloneWith(this.#props.dropTableNode, {
                cascade: true,
            }),
        });
    }
    toOperationNode() {
        return this.#props.executor.transformQuery(this.#props.dropTableNode, this.#props.queryId);
    }
    compile() {
        return this.#props.executor.compileQuery(this.toOperationNode(), this.#props.queryId);
    }
    async execute() {
        await this.#props.executor.executeQuery(this.compile(), this.#props.queryId);
    }
}
exports.DropTableBuilder = DropTableBuilder;
(0, prevent_await_js_1.preventAwait)(DropTableBuilder, "don't await DropTableBuilder instances directly. To execute the query you need to call `execute`");
