// kysely/dist/cjs/schema/drop-schema-builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DropSchemaBuilder = void 0;
const drop_schema_node_js_1 = require("../operation-node/drop-schema-node.js");
const prevent_await_js_1 = require("../util/prevent-await.js");
const object_utils_js_1 = require("../util/object-utils.js");
class DropSchemaBuilder {
    #props;
    constructor(props) {
        this.#props = (0, object_utils_js_1.freeze)(props);
    }
    ifExists() {
        return new DropSchemaBuilder({
            ...this.#props,
            dropSchemaNode: drop_schema_node_js_1.DropSchemaNode.cloneWith(this.#props.dropSchemaNode, {
                ifExists: true,
            }),
        });
    }
    cascade() {
        return new DropSchemaBuilder({
            ...this.#props,
            dropSchemaNode: drop_schema_node_js_1.DropSchemaNode.cloneWith(this.#props.dropSchemaNode, {
                cascade: true,
            }),
        });
    }
    toOperationNode() {
        return this.#props.executor.transformQuery(this.#props.dropSchemaNode, this.#props.queryId);
    }
    compile() {
        return this.#props.executor.compileQuery(this.toOperationNode(), this.#props.queryId);
    }
    async execute() {
        await this.#props.executor.executeQuery(this.compile(), this.#props.queryId);
    }
}
exports.DropSchemaBuilder = DropSchemaBuilder;
(0, prevent_await_js_1.preventAwait)(DropSchemaBuilder, "don't await DropSchemaBuilder instances directly. To execute the query you need to call `execute`");
