// kysely/dist/cjs/parser/reference-parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseColumnName = exports.parseAliasedStringReference = exports.parseStringReference = exports.parseReferenceExpression = exports.parseReferenceExpressionOrList = void 0;
const alias_node_js_1 = require("../operation-node/alias-node.js");
const column_node_js_1 = require("../operation-node/column-node.js");
const reference_node_js_1 = require("../operation-node/reference-node.js");
const table_node_js_1 = require("../operation-node/table-node.js");
const object_utils_js_1 = require("../util/object-utils.js");
const complex_expression_parser_js_1 = require("./complex-expression-parser.js");
const dynamic_reference_builder_js_1 = require("../dynamic/dynamic-reference-builder.js");
const identifier_node_js_1 = require("../operation-node/identifier-node.js");
function parseReferenceExpressionOrList(arg) {
    if ((0, object_utils_js_1.isReadonlyArray)(arg)) {
        return arg.map((it) => parseReferenceExpression(it));
    }
    else {
        return [parseReferenceExpression(arg)];
    }
}
exports.parseReferenceExpressionOrList = parseReferenceExpressionOrList;
function parseReferenceExpression(exp) {
    if ((0, object_utils_js_1.isString)(exp)) {
        return parseStringReference(exp);
    }
    else if ((0, dynamic_reference_builder_js_1.isDynamicReferenceBuilder)(exp)) {
        return exp.toOperationNode();
    }
    return (0, complex_expression_parser_js_1.parseComplexExpression)(exp);
}
exports.parseReferenceExpression = parseReferenceExpression;
function parseStringReference(ref) {
    const COLUMN_SEPARATOR = '.';
    if (ref.includes(COLUMN_SEPARATOR)) {
        const parts = ref.split(COLUMN_SEPARATOR).map(trim);
        if (parts.length === 3) {
            return parseStringReferenceWithTableAndSchema(parts);
        }
        else if (parts.length === 2) {
            return parseStringReferenceWithTable(parts);
        }
        else {
            throw new Error(`invalid column reference ${ref}`);
        }
    }
    else {
        return column_node_js_1.ColumnNode.create(ref);
    }
}
exports.parseStringReference = parseStringReference;
function parseAliasedStringReference(ref) {
    const ALIAS_SEPARATOR = ' as ';
    if (ref.includes(ALIAS_SEPARATOR)) {
        const [columnRef, alias] = ref.split(ALIAS_SEPARATOR).map(trim);
        return alias_node_js_1.AliasNode.create(parseStringReference(columnRef), identifier_node_js_1.IdentifierNode.create(alias));
    }
    else {
        return parseStringReference(ref);
    }
}
exports.parseAliasedStringReference = parseAliasedStringReference;
function parseColumnName(column) {
    return column_node_js_1.ColumnNode.create(column);
}
exports.parseColumnName = parseColumnName;
function parseStringReferenceWithTableAndSchema(parts) {
    const [schema, table, column] = parts;
    return reference_node_js_1.ReferenceNode.create(table_node_js_1.TableNode.createWithSchema(schema, table), column_node_js_1.ColumnNode.create(column));
}
function parseStringReferenceWithTable(parts) {
    const [table, column] = parts;
    return reference_node_js_1.ReferenceNode.create(table_node_js_1.TableNode.create(table), column_node_js_1.ColumnNode.create(column));
}
function trim(str) {
    return str.trim();
}
