// kysely/dist/cjs/dialect/mysql/mysql-introspector.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MysqlIntrospector = void 0;
const migrator_js_1 = require("../../migration/migrator.js");
const object_utils_js_1 = require("../../util/object-utils.js");
const sql_js_1 = require("../../raw-builder/sql.js");
class MysqlIntrospector {
    #db;
    constructor(db) {
        this.#db = db;
    }
    async getSchemas() {
        let rawSchemas = await this.#db
            .selectFrom('information_schema.schemata')
            .select('schema_name')
            .castTo()
            .execute();
        return rawSchemas.map((it) => ({ name: it.SCHEMA_NAME }));
    }
    async getTables(options = { withInternalKyselyTables: false }) {
        let query = this.#db
            .selectFrom('information_schema.columns')
            .select([
            'COLUMN_NAME',
            'COLUMN_DEFAULT',
            'TABLE_NAME',
            'TABLE_SCHEMA',
            'IS_NULLABLE',
            'DATA_TYPE',
            'EXTRA',
        ])
            .where('table_schema', '=', (0, sql_js_1.sql) `database()`)
            .orderBy('table_name')
            .orderBy('column_name')
            .castTo();
        if (!options.withInternalKyselyTables) {
            query = query
                .where('table_name', '!=', migrator_js_1.DEFAULT_MIGRATION_TABLE)
                .where('table_name', '!=', migrator_js_1.DEFAULT_MIGRATION_LOCK_TABLE);
        }
        const rawColumns = await query.execute();
        return this.#parseTableMetadata(rawColumns);
    }
    async getMetadata(options) {
        return {
            tables: await this.getTables(options),
        };
    }
    #parseTableMetadata(columns) {
        return columns.reduce((tables, it) => {
            let table = tables.find((tbl) => tbl.name === it.TABLE_NAME);
            if (!table) {
                table = (0, object_utils_js_1.freeze)({
                    name: it.TABLE_NAME,
                    schema: it.TABLE_SCHEMA,
                    columns: [],
                });
                tables.push(table);
            }
            table.columns.push((0, object_utils_js_1.freeze)({
                name: it.COLUMN_NAME,
                dataType: it.DATA_TYPE,
                isNullable: it.IS_NULLABLE === 'YES',
                isAutoIncrementing: it.EXTRA.toLowerCase().includes('auto_increment'),
                hasDefaultValue: it.COLUMN_DEFAULT !== null,
            }));
            return tables;
        }, []);
    }
}
exports.MysqlIntrospector = MysqlIntrospector;
