// kysely/dist/cjs/dialect/sqlite/sqlite-introspector.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SqliteIntrospector = void 0;
const migrator_js_1 = require("../../migration/migrator.js");
const sql_js_1 = require("../../raw-builder/sql.js");
class SqliteIntrospector {
    #db;
    constructor(db) {
        this.#db = db;
    }
    async getSchemas() {
        // Sqlite doesn't support schemas.
        return [];
    }
    async getTables(options = { withInternalKyselyTables: false }) {
        let query = this.#db
            .selectFrom('sqlite_schema')
            .where('type', '=', 'table')
            .where('name', 'not like', 'sqlite_%')
            .select('name')
            .orderBy('name')
            .castTo();
        if (!options.withInternalKyselyTables) {
            query = query
                .where('name', '!=', migrator_js_1.DEFAULT_MIGRATION_TABLE)
                .where('name', '!=', migrator_js_1.DEFAULT_MIGRATION_LOCK_TABLE);
        }
        const tables = await query.execute();
        return Promise.all(tables.map(({ name }) => this.#getTableMetadata(name)));
    }
    async getMetadata(options) {
        return {
            tables: await this.getTables(options),
        };
    }
    async #getTableMetadata(table) {
        const db = this.#db;
        // Get the SQL that was used to create the table.
        const createSql = await db
            .selectFrom('sqlite_master')
            .where('name', '=', table)
            .select('sql')
            .castTo()
            .execute();
        // Try to find the name of the column that has `autoincrement` 🤦
        const autoIncrementCol = createSql[0]?.sql
            ?.split(/[\(\),]/)
            ?.find((it) => it.toLowerCase().includes('autoincrement'))
            ?.split(/\s+/)?.[0]
            ?.replace(/["`]/g, '');
        const columns = await db
            .selectFrom((0, sql_js_1.sql) `pragma_table_info(${table})`.as('table_info'))
            .select(['name', 'type', 'notnull', 'dflt_value'])
            .orderBy('name')
            .execute();
        return {
            name: table,
            columns: columns.map((col) => ({
                name: col.name,
                dataType: col.type,
                isNullable: !col.notnull,
                isAutoIncrementing: col.name === autoIncrementCol,
                hasDefaultValue: col.dflt_value != null,
            })),
        };
    }
}
exports.SqliteIntrospector = SqliteIntrospector;
