// kysely/dist/cjs/raw-builder/sql.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sql = void 0;
const identifier_node_js_1 = require("../operation-node/identifier-node.js");
const raw_node_js_1 = require("../operation-node/raw-node.js");
const value_node_js_1 = require("../operation-node/value-node.js");
const reference_parser_js_1 = require("../parser/reference-parser.js");
const table_parser_js_1 = require("../parser/table-parser.js");
const value_parser_js_1 = require("../parser/value-parser.js");
const query_id_js_1 = require("../util/query-id.js");
const raw_builder_js_1 = require("./raw-builder.js");
exports.sql = Object.assign((sqlFragments, ...parameters) => {
    return new raw_builder_js_1.RawBuilder({
        queryId: (0, query_id_js_1.createQueryId)(),
        rawNode: raw_node_js_1.RawNode.create(sqlFragments, parameters?.map(value_parser_js_1.parseValueExpression) ?? []),
    });
}, {
    ref(columnReference) {
        return new raw_builder_js_1.RawBuilder({
            queryId: (0, query_id_js_1.createQueryId)(),
            rawNode: raw_node_js_1.RawNode.createWithChild((0, reference_parser_js_1.parseStringReference)(columnReference)),
        });
    },
    value(value) {
        return new raw_builder_js_1.RawBuilder({
            queryId: (0, query_id_js_1.createQueryId)(),
            rawNode: raw_node_js_1.RawNode.createWithChild((0, value_parser_js_1.parseValueExpression)(value)),
        });
    },
    table(tableReference) {
        return new raw_builder_js_1.RawBuilder({
            queryId: (0, query_id_js_1.createQueryId)(),
            rawNode: raw_node_js_1.RawNode.createWithChild((0, table_parser_js_1.parseTable)(tableReference)),
        });
    },
    id(...ids) {
        const fragments = new Array(ids.length + 1).fill('.');
        fragments[0] = '';
        fragments[fragments.length - 1] = '';
        return new raw_builder_js_1.RawBuilder({
            queryId: (0, query_id_js_1.createQueryId)(),
            rawNode: raw_node_js_1.RawNode.create(fragments, ids.map(identifier_node_js_1.IdentifierNode.create)),
        });
    },
    literal(value) {
        return new raw_builder_js_1.RawBuilder({
            queryId: (0, query_id_js_1.createQueryId)(),
            rawNode: raw_node_js_1.RawNode.createWithChild(value_node_js_1.ValueNode.createImmediate(value)),
        });
    },
    raw(sql) {
        return new raw_builder_js_1.RawBuilder({
            queryId: (0, query_id_js_1.createQueryId)(),
            rawNode: raw_node_js_1.RawNode.createWithSql(sql),
        });
    },
    join(array, separator = (0, exports.sql) `, `) {
        const nodes = new Array(2 * array.length - 1);
        const sep = separator.toOperationNode();
        for (let i = 0; i < array.length; ++i) {
            nodes[2 * i] = (0, value_parser_js_1.parseValueExpression)(array[i]);
            if (i !== array.length - 1) {
                nodes[2 * i + 1] = sep;
            }
        }
        return new raw_builder_js_1.RawBuilder({
            queryId: (0, query_id_js_1.createQueryId)(),
            rawNode: raw_node_js_1.RawNode.createWithChildren(nodes),
        });
    },
});
