// kysely/dist/esm/query-builder/join-builder.js
/// <reference types="./join-builder.d.ts" />
import { JoinNode } from '../operation-node/join-node.js';
import { RawNode } from '../operation-node/raw-node.js';
import { parseExistFilter, parseNotExistFilter, parseOnFilter, parseReferenceFilter, } from '../parser/filter-parser.js';
import { freeze } from '../util/object-utils.js';
import { preventAwait } from '../util/prevent-await.js';
export class JoinBuilder {
    #props;
    constructor(props) {
        this.#props = freeze(props);
    }
    on(...args) {
        return new JoinBuilder({
            ...this.#props,
            joinNode: JoinNode.cloneWithOn(this.#props.joinNode, parseOnFilter(args)),
        });
    }
    orOn(...args) {
        return new JoinBuilder({
            ...this.#props,
            joinNode: JoinNode.cloneWithOrOn(this.#props.joinNode, parseOnFilter(args)),
        });
    }
    /**
     * Just like {@link WhereInterface.whereRef} but adds an item to the join's
     * `on` clause instead.
     *
     * See {@link WhereInterface.whereRef} for documentation and examples.
     */
    onRef(lhs, op, rhs) {
        return new JoinBuilder({
            ...this.#props,
            joinNode: JoinNode.cloneWithOn(this.#props.joinNode, parseReferenceFilter(lhs, op, rhs)),
        });
    }
    /**
     * Just like {@link WhereInterface.orWhereRef} but adds an item to the join's
     * `on` clause instead.
     *
     * See {@link WhereInterface.orWhereRef} for documentation and examples.
     */
    orOnRef(lhs, op, rhs) {
        return new JoinBuilder({
            ...this.#props,
            joinNode: JoinNode.cloneWithOrOn(this.#props.joinNode, parseReferenceFilter(lhs, op, rhs)),
        });
    }
    /**
     * Just like {@link WhereInterface.whereExists} but adds an item to the join's
     * `on` clause instead.
     *
     * See {@link WhereInterface.whereExists} for documentation and examples.
     */
    onExists(arg) {
        return new JoinBuilder({
            ...this.#props,
            joinNode: JoinNode.cloneWithOn(this.#props.joinNode, parseExistFilter(arg)),
        });
    }
    /**
     * Just like {@link WhereInterface.whereNotExists} but adds an item to the join's
     * `on` clause instead.
     *
     * See {@link WhereInterface.whereNotExists} for documentation and examples.
     */
    onNotExists(arg) {
        return new JoinBuilder({
            ...this.#props,
            joinNode: JoinNode.cloneWithOn(this.#props.joinNode, parseNotExistFilter(arg)),
        });
    }
    /**
     * Just like {@link WhereInterface.orWhereExists} but adds an item to the join's
     * `on` clause instead.
     *
     * See {@link WhereInterface.orWhereExists} for documentation and examples.
     */
    orOnExists(arg) {
        return new JoinBuilder({
            ...this.#props,
            joinNode: JoinNode.cloneWithOrOn(this.#props.joinNode, parseExistFilter(arg)),
        });
    }
    /**
     * Just like {@link WhereInterface.orWhereNotExists} but adds an item to the join's
     * `on` clause instead.
     *
     * See {@link WhereInterface.orWhereNotExists} for documentation and examples.
     */
    orOnNotExists(arg) {
        return new JoinBuilder({
            ...this.#props,
            joinNode: JoinNode.cloneWithOrOn(this.#props.joinNode, parseNotExistFilter(arg)),
        });
    }
    /**
     * Adds `on true`.
     */
    onTrue() {
        return new JoinBuilder({
            ...this.#props,
            joinNode: JoinNode.cloneWithOn(this.#props.joinNode, RawNode.createWithSql('true')),
        });
    }
    toOperationNode() {
        return this.#props.joinNode;
    }
}
preventAwait(JoinBuilder, "don't await JoinBuilder instances. They are never executed directly and are always just a part of a query.");
