// kysely/dist/cjs/util/log.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Log = exports.LOG_LEVELS = void 0;
const object_utils_js_1 = require("./object-utils.js");
exports.LOG_LEVELS = (0, object_utils_js_1.freeze)(['query', 'error']);
class Log {
    #levels;
    #logger;
    constructor(config) {
        if ((0, object_utils_js_1.isFunction)(config)) {
            this.#logger = config;
            this.#levels = (0, object_utils_js_1.freeze)({
                query: true,
                error: true,
            });
        }
        else {
            this.#logger = defaultLogger;
            this.#levels = (0, object_utils_js_1.freeze)({
                query: config.includes('query'),
                error: config.includes('error'),
            });
        }
    }
    isLevelEnabled(level) {
        return this.#levels[level];
    }
    query(getEvent) {
        if (this.#levels.query) {
            this.#logger(getEvent());
        }
    }
    error(getEvent) {
        if (this.#levels.error) {
            this.#logger(getEvent());
        }
    }
}
exports.Log = Log;
function defaultLogger(event) {
    if (event.level === 'query') {
        console.log(`kysely:query: ${event.query.sql}`);
        console.log(`kysely:query: duration: ${event.queryDurationMillis.toFixed(1)}ms`);
    }
    else if (event.level === 'error') {
        if (event.error instanceof Error) {
            console.error(`kysely:error: ${event.error.stack ?? event.error.message}`);
        }
        else {
            console.error(`kysely:error: ${event}`);
        }
    }
}
