// kysely/dist/esm/util/log.js
/// <reference types="./log.d.ts" />
import { freeze, isFunction } from './object-utils.js';
export const LOG_LEVELS = freeze(['query', 'error']);
export class Log {
    #levels;
    #logger;
    constructor(config) {
        if (isFunction(config)) {
            this.#logger = config;
            this.#levels = freeze({
                query: true,
                error: true,
            });
        }
        else {
            this.#logger = defaultLogger;
            this.#levels = freeze({
                query: config.includes('query'),
                error: config.includes('error'),
            });
        }
    }
    isLevelEnabled(level) {
        return this.#levels[level];
    }
    query(getEvent) {
        if (this.#levels.query) {
            this.#logger(getEvent());
        }
    }
    error(getEvent) {
        if (this.#levels.error) {
            this.#logger(getEvent());
        }
    }
}
function defaultLogger(event) {
    if (event.level === 'query') {
        console.log(`kysely:query: ${event.query.sql}`);
        console.log(`kysely:query: duration: ${event.queryDurationMillis.toFixed(1)}ms`);
    }
    else if (event.level === 'error') {
        if (event.error instanceof Error) {
            console.error(`kysely:error: ${event.error.stack ?? event.error.message}`);
        }
        else {
            console.error(`kysely:error: ${event}`);
        }
    }
}
