// kysely/dist/esm/schema/drop-index-builder.js
/// <reference types="./drop-index-builder.d.ts" />
import { DropIndexNode } from '../operation-node/drop-index-node.js';
import { preventAwait } from '../util/prevent-await.js';
import { parseTable } from '../parser/table-parser.js';
import { freeze } from '../util/object-utils.js';
export class DropIndexBuilder {
    #props;
    constructor(props) {
        this.#props = freeze(props);
    }
    /**
     * Specifies the table the index was created for. This is not needed
     * in all dialects.
     */
    on(table) {
        return new DropIndexBuilder({
            ...this.#props,
            dropIndexNode: DropIndexNode.cloneWith(this.#props.dropIndexNode, {
                table: parseTable(table),
            }),
        });
    }
    ifExists() {
        return new DropIndexBuilder({
            ...this.#props,
            dropIndexNode: DropIndexNode.cloneWith(this.#props.dropIndexNode, {
                ifExists: true,
            }),
        });
    }
    cascade() {
        return new DropIndexBuilder({
            ...this.#props,
            dropIndexNode: DropIndexNode.cloneWith(this.#props.dropIndexNode, {
                cascade: true,
            }),
        });
    }
    toOperationNode() {
        return this.#props.executor.transformQuery(this.#props.dropIndexNode, this.#props.queryId);
    }
    compile() {
        return this.#props.executor.compileQuery(this.toOperationNode(), this.#props.queryId);
    }
    async execute() {
        await this.#props.executor.executeQuery(this.compile(), this.#props.queryId);
    }
}
preventAwait(DropIndexBuilder, "don't await DropIndexBuilder instances directly. To execute the query you need to call `execute`");
