// kysely/dist/esm/schema/drop-schema-builder.js
/// <reference types="./drop-schema-builder.d.ts" />
import { DropSchemaNode } from '../operation-node/drop-schema-node.js';
import { preventAwait } from '../util/prevent-await.js';
import { freeze } from '../util/object-utils.js';
export class DropSchemaBuilder {
    #props;
    constructor(props) {
        this.#props = freeze(props);
    }
    ifExists() {
        return new DropSchemaBuilder({
            ...this.#props,
            dropSchemaNode: DropSchemaNode.cloneWith(this.#props.dropSchemaNode, {
                ifExists: true,
            }),
        });
    }
    cascade() {
        return new DropSchemaBuilder({
            ...this.#props,
            dropSchemaNode: DropSchemaNode.cloneWith(this.#props.dropSchemaNode, {
                cascade: true,
            }),
        });
    }
    toOperationNode() {
        return this.#props.executor.transformQuery(this.#props.dropSchemaNode, this.#props.queryId);
    }
    compile() {
        return this.#props.executor.compileQuery(this.toOperationNode(), this.#props.queryId);
    }
    async execute() {
        await this.#props.executor.executeQuery(this.compile(), this.#props.queryId);
    }
}
preventAwait(DropSchemaBuilder, "don't await DropSchemaBuilder instances directly. To execute the query you need to call `execute`");
