// kysely/dist/esm/schema/drop-table-builder.js
/// <reference types="./drop-table-builder.d.ts" />
import { DropTableNode } from '../operation-node/drop-table-node.js';
import { preventAwait } from '../util/prevent-await.js';
import { freeze } from '../util/object-utils.js';
export class DropTableBuilder {
    #props;
    constructor(props) {
        this.#props = freeze(props);
    }
    ifExists() {
        return new DropTableBuilder({
            ...this.#props,
            dropTableNode: DropTableNode.cloneWith(this.#props.dropTableNode, {
                ifExists: true,
            }),
        });
    }
    cascade() {
        return new DropTableBuilder({
            ...this.#props,
            dropTableNode: DropTableNode.cloneWith(this.#props.dropTableNode, {
                cascade: true,
            }),
        });
    }
    toOperationNode() {
        return this.#props.executor.transformQuery(this.#props.dropTableNode, this.#props.queryId);
    }
    compile() {
        return this.#props.executor.compileQuery(this.toOperationNode(), this.#props.queryId);
    }
    async execute() {
        await this.#props.executor.executeQuery(this.compile(), this.#props.queryId);
    }
}
preventAwait(DropTableBuilder, "don't await DropTableBuilder instances directly. To execute the query you need to call `execute`");
