// kysely/dist/cjs/operation-node/on-conflict-node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OnConflictNode = void 0;
const object_utils_js_1 = require("../util/object-utils.js");
const where_node_js_1 = require("./where-node.js");
/**
 * @internal
 */
exports.OnConflictNode = (0, object_utils_js_1.freeze)({
    is(node) {
        return node.kind === 'OnConflictNode';
    },
    create() {
        return (0, object_utils_js_1.freeze)({
            kind: 'OnConflictNode',
        });
    },
    cloneWith(node, props) {
        return (0, object_utils_js_1.freeze)({
            ...node,
            ...props,
        });
    },
    cloneWithIndexWhere(node, filter) {
        return (0, object_utils_js_1.freeze)({
            ...node,
            indexWhere: node.indexWhere
                ? where_node_js_1.WhereNode.cloneWithFilter(node.indexWhere, 'And', filter)
                : where_node_js_1.WhereNode.create(filter),
        });
    },
    cloneWithIndexOrWhere(node, filter) {
        return (0, object_utils_js_1.freeze)({
            ...node,
            indexWhere: node.indexWhere
                ? where_node_js_1.WhereNode.cloneWithFilter(node.indexWhere, 'Or', filter)
                : where_node_js_1.WhereNode.create(filter),
        });
    },
    cloneWithUpdateWhere(node, filter) {
        return (0, object_utils_js_1.freeze)({
            ...node,
            updateWhere: node.updateWhere
                ? where_node_js_1.WhereNode.cloneWithFilter(node.updateWhere, 'And', filter)
                : where_node_js_1.WhereNode.create(filter),
        });
    },
    cloneWithUpdateOrWhere(node, filter) {
        return (0, object_utils_js_1.freeze)({
            ...node,
            updateWhere: node.updateWhere
                ? where_node_js_1.WhereNode.cloneWithFilter(node.updateWhere, 'Or', filter)
                : where_node_js_1.WhereNode.create(filter),
        });
    },
});
