// kysely/dist/cjs/operation-node/select-query-node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SelectQueryNode = void 0;
const object_utils_js_1 = require("../util/object-utils.js");
const from_node_js_1 = require("./from-node.js");
const group_by_node_js_1 = require("./group-by-node.js");
const having_node_js_1 = require("./having-node.js");
const order_by_node_js_1 = require("./order-by-node.js");
/**
 * @internal
 */
exports.SelectQueryNode = (0, object_utils_js_1.freeze)({
    is(node) {
        return node.kind === 'SelectQueryNode';
    },
    create(fromItems, withNode) {
        return (0, object_utils_js_1.freeze)({
            kind: 'SelectQueryNode',
            from: from_node_js_1.FromNode.create(fromItems),
            ...(withNode && { with: withNode }),
        });
    },
    cloneWithSelections(select, selections) {
        return (0, object_utils_js_1.freeze)({
            ...select,
            selections: select.selections
                ? (0, object_utils_js_1.freeze)([...select.selections, ...selections])
                : (0, object_utils_js_1.freeze)(selections),
        });
    },
    cloneWithDistinctOnSelections(select, selections) {
        return (0, object_utils_js_1.freeze)({
            ...select,
            distinctOnSelections: select.distinctOnSelections
                ? (0, object_utils_js_1.freeze)([...select.distinctOnSelections, ...selections])
                : (0, object_utils_js_1.freeze)(selections),
        });
    },
    cloneWithFrontModifier(select, modifier) {
        return (0, object_utils_js_1.freeze)({
            ...select,
            frontModifiers: select.frontModifiers
                ? (0, object_utils_js_1.freeze)([...select.frontModifiers, modifier])
                : (0, object_utils_js_1.freeze)([modifier]),
        });
    },
    cloneWithEndModifier(select, modifier) {
        return (0, object_utils_js_1.freeze)({
            ...select,
            endModifiers: select.endModifiers
                ? (0, object_utils_js_1.freeze)([...select.endModifiers, modifier])
                : (0, object_utils_js_1.freeze)([modifier]),
        });
    },
    cloneWithOrderByItem(selectNode, item) {
        return (0, object_utils_js_1.freeze)({
            ...selectNode,
            orderBy: selectNode.orderBy
                ? order_by_node_js_1.OrderByNode.cloneWithItem(selectNode.orderBy, item)
                : order_by_node_js_1.OrderByNode.create(item),
        });
    },
    cloneWithGroupByItems(selectNode, items) {
        return (0, object_utils_js_1.freeze)({
            ...selectNode,
            groupBy: selectNode.groupBy
                ? group_by_node_js_1.GroupByNode.cloneWithItems(selectNode.groupBy, items)
                : group_by_node_js_1.GroupByNode.create(items),
        });
    },
    cloneWithLimit(selectNode, limit) {
        return (0, object_utils_js_1.freeze)({
            ...selectNode,
            limit,
        });
    },
    cloneWithOffset(selectNode, offset) {
        return (0, object_utils_js_1.freeze)({
            ...selectNode,
            offset,
        });
    },
    cloneWithHaving(selectNode, filter) {
        return (0, object_utils_js_1.freeze)({
            ...selectNode,
            having: selectNode.having
                ? having_node_js_1.HavingNode.cloneWithFilter(selectNode.having, 'And', filter)
                : having_node_js_1.HavingNode.create(filter),
        });
    },
    cloneWithOrHaving(selectNode, filter) {
        return (0, object_utils_js_1.freeze)({
            ...selectNode,
            having: selectNode.having
                ? having_node_js_1.HavingNode.cloneWithFilter(selectNode.having, 'Or', filter)
                : having_node_js_1.HavingNode.create(filter),
        });
    },
    cloneWithSetOperation(selectNode, setOperation) {
        return (0, object_utils_js_1.freeze)({
            ...selectNode,
            setOperations: selectNode.setOperations
                ? (0, object_utils_js_1.freeze)([...selectNode.setOperations, setOperation])
                : (0, object_utils_js_1.freeze)([setOperation]),
        });
    },
    cloneWithExplain(selectNode, explain) {
        return (0, object_utils_js_1.freeze)({
            ...selectNode,
            explain,
        });
    },
});
